Sources/Purchasing/Purchases/PurchasesOrchestrator.swift

```diff
@@ -823,19 +830,27 @@ private extension PurchasesOrchestrator {
     /// - Parameter restored: whether the transaction state was `.restored` instead of `.purchased`.
     private func purchaseSource(
         for productIdentifier: String,
+        transaction: StoreTransaction,
         restored: Bool
     ) -> PurchaseSource {
         let initiationSource: ProductRequestData.InitiationSource = {
             // Having a purchase completed callback implies that the transation comes from an explicit call
             // to `purchase()` instead of a StoreKit transaction notification.
-            let hasPurchaseCallback = self.purchaseCompleteCallbacksByProductID.value.keys.contains(productIdentifier)
+            // As long as the transaction date is _after_ the completion block was added.
+            let completionBlockDate = self.purchaseCompleteCallbacksByProductID.value[productIdentifier]?.date
+
+            switch (completionBlockDate, restored) {
+            case let (.some(completionBlockDate), false):
+                if transaction.purchaseDate > completionBlockDate {
+                    return .purchase
```

Comment:
we shouldn't just change the `initiation_source`, we should entirely not use this callback if the transaction is older than when you started the purchase. 

It's somewhat delicate in that we still need to handle transactions with no dates (like for user cancels), but if we can definitely say "this transaction isn't the one I started the purchase for" then the best path would be to just not call completion for it, right?

Reply:
Oooooh of course. Let's talk through together this tomorrow if you can 🙏🏻